drivers/blk/sdmmc/sdmmc_protocol/sdmmc.rs

```diff
@@ -0,0 +1,405 @@
+use core::{future::Future, pin::Pin, task::{Context, Poll, Waker}};
+
+use sdmmc_constant::{MMC_CMD_READ_MULTIPLE_BLOCK, MMC_CMD_READ_SINGLE_BLOCK, MMC_CMD_STOP_TRANSMISSION, MMC_CMD_WRITE_MULTIPLE_BLOCK, MMC_CMD_WRITE_SINGLE_BLOCK};
+mod sdmmc_constant;
+pub struct SdmmcCmd {
+    pub cmdidx: u32,
+    pub resp_type: u32,
+    pub cmdarg: u32,
+}
+
+pub struct MmcData {
+    // The size of the block(sector size), for sdcard should almost always be 512
+    pub blocksize: u32,
+    // Number of blocks to transfer
+    pub blockcnt: u32,
+    pub flags: MmcDataFlag,
+    pub addr: u64,
+}
+
+pub enum MmcDataFlag {
+    SdmmcDataRead,
+    SdmmcDataWrite,
+}
+
+pub enum SdmmcHalError {
+    // Error for result not ready yet
+    EBUSY,
+    ETIMEDOUT,
+    EINVAL,
+    EIO,
+    ENOTIMPLEMENTED,
+    // This error should not be triggered unless there are bugs in program
+    EUNDEFINED,
+    // The block transfer succeed, but fail to stop the read/write process
+    ESTOPCMD,
+}
+
+// Interrupt related define
+#[repr(u32)]  // Ensures the enum variants are stored as 32-bit integers
+pub enum InterruptType {
+    Success = 0b0001,    // 1st bit
+    Error   = 0b0010,    // 2nd bit
+    SDIO    = 0b0100,    // 3rd bit
+    // You can add more flags as needed
+}
+
+// Define the MMC response flags
+const MMC_RSP_PRESENT: u32 = 1 << 0;
+const MMC_RSP_136: u32 = 1 << 1;       // 136-bit response
+const MMC_RSP_CRC: u32 = 1 << 2;       // Expect valid CRC
+const MMC_RSP_BUSY: u32 = 1 << 3;      // Card may send busy
+const MMC_RSP_OPCODE: u32 = 1 << 4;    // Response contains opcode
+
+// Define the MMC response types
+pub const MMC_RSP_NONE: u32 = 0;
+pub const MMC_RSP_R1: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE;
+pub const MMC_RSP_R1B: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE | MMC_RSP_BUSY;
+pub const MMC_RSP_R2: u32 = MMC_RSP_PRESENT | MMC_RSP_136 | MMC_RSP_CRC;
+pub const MMC_RSP_R3: u32 = MMC_RSP_PRESENT;
+pub const MMC_RSP_R4: u32 = MMC_RSP_PRESENT;
+pub const MMC_RSP_R5: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE;
+pub const MMC_RSP_R6: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE;
+pub const MMC_RSP_R7: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE;
+
+
+/// Program async Rust can be very dangerous if you do not know what is happening understand the hood
+/// Power up and power off cannot be properly implemented if I do not have access to control gpio/ regulator and timer
+pub trait SdmmcHardware {
```

Comment:
It's probably so that a concrete implementation can implement a susbet of the full functionality of the trait, right?

Not clear if this is something that we actually want or not...